File: app/src/main/java/com/example/pam_project/networking/RetrofitUtils.java
package com.example.pam_project.networking;

import io.reactivex.Single;
import retrofit2.Response;

public final class RetrofitUtils {
    public static <T> Single<T> performRequest(final Single<Response<T>> request) {
        return request.onErrorResumeNext(Single::error)
                .map(RetrofitUtils::unwrapResponse);
    }

    public static <T> T unwrapResponse(final Response<T> response) {
        if (RetrofitUtils.isRequestFailed(response)) {
            throw new IllegalArgumentException();
        }

        return response.body();
    }

    private static boolean isRequestFailed(final Response response) {
        return !response.isSuccessful() || isErrorCode(response.code());
    }

    private static boolean isErrorCode(final int statusCode) {
        return statusCode >= 400 && statusCode < 600;
    }
}

File: app/src/main/java/com/example/pam_project/networking/RetrofitUtils.kt
package com.example.pam_project.networking

import io.reactivex.Single
import retrofit2.Response

object RetrofitUtils {
    fun <T> performRequest(request: Single<Response<T>?>?): Single<T> {
        return request!!.onErrorResumeNext { exception: Throwable? -> Single.error(exception) }
                .map { obj: Response<T>? -> unwrapResponse() }
    }

    fun <T> unwrapResponse(response: Response<T>): T? {
        require(!isRequestFailed(response))
        return response.body()
    }

    private fun <T> isRequestFailed(response: Response<T>): Boolean {
        return !response.isSuccessful || isErrorCode(response.code())
    }

    private fun isErrorCode(statusCode: Int): Boolean {
        return statusCode >= 400 && statusCode < 600
    }
}